standards/core/rln-contract.md

```diff
@@ -180,6 +204,17 @@ and the membership set SHOULD be migrated.
 
 ## Implementation Suggestions
 
+### Choosing Which _Expired_ Memberships to Overwrite
+
+When registering a new membership, the smart contract may need to decide which _Expired_ memberships to overwrite.
+The criteria for this selection can vary depending on the implementation.
+
+Key considerations include:
+- To minimize gas costs, it's better to overwrite a single high-rate membership rather than multiple low-rate ones.
+- To encourage timely withdrawals, it's better to overwrite memberships that have been _Expired_ for a long time.
```

Comment:
Do we agree @richard-ramos that for now, we don't include this feature in the smart contract?

Reply:
The way it works right now is that you can:
- Specify the memberships you want to expire
- If you don't specify:
- - If there's enough space in the tree / enough rate limit: create the membership
- - Otherwise, attempt to free it.

Since for any option ideally the client should call `estimateGas` to determine which option is cheaper, and the smart contract's been already modified to support this scenario, we can keep it. Unless we want to remove the tracking of the membership age in the smart contract completely? If so, then, cool. It means less gas for the user, and nwaku can track the age of the memberships then.

Reply:
> nwaku can track the age of the membership

Are you saying that nwaku would need to listen to emit events from the smart contract? I think we definitely do not want that as the whole point of onchain tree was to not have to load all events.

However, can the solution be:
- membership and some age information are stored in contract
- BUT they are not ordered

Meaning that it's cheaper to insert or expire (no ordering at insertion). But it's possible to get all data from the contract API (no tracing of emitted events).
The ordering can then be done by the applicaiton (nwaku, jswaku) etc

Reply:
Sure! no need to use events. There's already a function that would let you iterate over all the memberships registered. Also perhaps in the future we might want to integrate with the graph and obtain this information from there too. 

I'll propose the changes on a separate branch from https://github.com/waku-org/waku-rlnv2-contract/pull/13 so it's possible to compare both options. I do think not having to track the ordering in the contract will be very beneficial due to reduced costs